solr/contrib/ltr/src/java/org/apache/solr/ltr/LTRRescorer.java

```diff
@@ -166,64 +186,77 @@ public void scoreFeatures(IndexSearcher indexSearcher, TopDocs firstPassTopDocs,
         docBase = readerContext.docBase;
         scorer = modelWeight.scorer(readerContext);
       }
-      // Scorer for a LTRScoringQuery.ModelWeight should never be null since we always have to
-      // call score
-      // even if no feature scorers match, since a model might use that info to
-      // return a
-      // non-zero score. Same applies for the case of advancing a LTRScoringQuery.ModelWeight.ModelScorer
-      // past the target
-      // doc since the model algorithm still needs to compute a potentially
-      // non-zero score from blank features.
-      assert (scorer != null);
-      final int targetDoc = docID - docBase;
-      scorer.docID();
-      scorer.iterator().advance(targetDoc);
-
-      scorer.getDocInfo().setOriginalDocScore(hit.score);
-      hit.score = scorer.score();
-      if (hitUpto < topN) {
-        reranked[hitUpto] = hit;
-        // if the heap is not full, maybe I want to log the features for this
-        // document
+      scoreSingleHit(indexSearcher, topN, modelWeight, docBase, hitUpto, hit, docID, scoringQuery, scorer, reranked);
```

Comment:
9/n observation and suggestions:
* very elegant factoring out of methods for use by `LTRInterleavingRescorer`
* the `rerank` method already takes a `searcher` and so it could determine its own `leaves` from that
* `this.scoringQuery` being passed to the `scoreSingleHit` method as an argument (rather than it using the `this.scoringQuery` directly) is very subtle. it is of course present as an argument because `LTRInterleavingRescorer` will passing `rerankingQueries[i]` for that argument. the subtlety could be removed by making `scoreSingleHit` a static method.

https://github.com/cpoerschke/lucene-solr/commit/16512dbddbabe12ca5a2a26a9180ceeaae62cea2

Reply:
Brilliant observation!